soc/arm/st_stm32/stm32wl/power.c

```diff
@@ -76,20 +86,34 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
 /* Handle SOC specific activity after Low Power Mode Exit */
 __weak void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
 {
-	ARG_UNUSED(substate_id);
-
 	switch (state) {
 	case PM_STATE_SUSPEND_TO_IDLE:
 		LL_LPM_DisableSleepOnExit();
 		LL_LPM_EnableSleep();
+		switch (substate_id) {
+		case 1: /* STOP0 */
+			break;
+		case 2: /* STOP1 */
+			break;
+		case 3: /* STOP2 */
+			/* Reinitialize peripheral functionality */
+			DT_FOREACH_STATUS_OKAY(st_stm32_adc,    STM32_ADC_REINIT);
+			DT_FOREACH_STATUS_OKAY(st_stm32_i2c_v2, STM32_I2C_REINIT);
+			DT_FOREACH_STATUS_OKAY(st_stm32_spi,    STM32_SPI_REINIT);
+			DT_FOREACH_STATUS_OKAY(st_stm32_usart,  STM32_USART_REINIT);
```

Comment:
definitely -1 on calling driver code from soc/. we need a mechanism so that drivers can subscribe to system power state changes and act accordingly.

Reply:
Yes this is messy, I agree.
I'm currently replacing it with a callback similar to pm_action_cb in the pm_device struct.

Reply:
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/pm/pm.h#L83-L91

PM subsystem send notifications when entering/leaving a power state, you can subscribe to receive them.

Reply:
I saw the `pm_notifier_` functions, but that unfortunately only gives high-level PM state information via the `pm_notifier` struct, i.e. `pm_state` enum. We need access to the `substate_id` to determine if we're coming out of e.g. STOP2 where we must reinit the peripheral, or STOP1 were we shouldn't.

My approach right now is still adding a `pm_reinit_cb` to the `pm_device` struct, which will be called via its own function inside `pm.c` `pm_resume_devices` and `pm_suspend_devices`, where `pm_resume_devices` is also called, but independent of `CONFIG_PM_DEVICE` and `CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE`.

Looking at the notifier stuff now, I suppose I could also extend the `pm_notifier` struct with `substate_id`, would you prefer this approach to the one above?

Trying to hammer this out in between the EOSS in-person meetings and talks, without losing too much mental context from the first two. :)

Reply:
I think the approach here is the problem ... Are you trying to suspend / resume devices manually because the subsystem is not doing it for  your state ? If that is the case I think the right thing is make this selection of which power state triggers device pm more generic (using device tree as you mentioned). In that case you can query the system state in your `pm_action_cb`. In this case when a device is asked for suspend you can see what is the next state the SoC is going and when it is asked to resume you can see what is the state that the SoC is leaving.

Reply:
Sorry, I must not have explained it clearly enough.

So the issue I'm trying to resolve here is in System PM. Issue example:

1. STM32WL SoC goes into STOP2. suspend-to-idle System PM low power mode
2. Some wakeup event wakes the SoC and CPU
3. The I2C register contents and clock. configuration have been lost, due to the way the STOP2 SoC low power mode works. These must now be reinitialized for I2C to function again.

My proposed `pm_reinit_cb` solution above merely piggybacks off of the `pm_device` struct, which is still present, even when PM Device is disabled due to the way the PM Device init macro in the driver works.

Are you still in Prague this Friday? I feel a 5 minute chat would be much more productive than this back and forth, should you have a minute :)

Reply:
> Sorry, I must not have explained it clearly enough.
> 
> So the issue I'm trying to resolve here is in System PM. Issue example:

Don't need to sorry at all :)

> 
> 1. STM32WL SoC goes into STOP2. suspend-to-idle System PM low power mode
> 2. Some wakeup event wakes the SoC and CPU

In this point I see two places where you can hook this code (both requires changes)

1) With the proposed changes to tell the system which states devices should be suspended / resumed, the pm subsystem can call `pm_device_action_run(dev, PM_DEVICE_ACTION_RESUME);` and you can check in the callback you can check if the system is returning from STOP2. This option assumes you have `CONFIG_PM_DEVICE` enabled.

2) Add more notifications ... 
```
pm_state_notify(true); /* entering STOP2 */
pm_state_set(STOP2);
pm_state_notify(false); /* Leaving STOP2 */
pm_state_exit_pos_ops(STOP2);
pm_state_notify(true); /* entering ACTIVE */

```

> 3. The I2C register contents and clock. configuration have been lost, due to the way the STOP2 SoC low power mode works. These must now be reinitialized for I2C to function again.
> 
> My proposed `pm_reinit_cb` solution above merely piggybacks off of the `pm_device` struct, which is still present, even when PM Device is disabled due to the way the PM Device init macro in the driver works.
> 
> Are you still in Prague this Friday? I feel a 5 minute chat would be much more productive than this back and forth, should you have a minute :)

Reply:
1. Again, this depends on a very uncommon and explicit configuration combination of `CONFIG_PM_DEVICE` and `!CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE`; not to mention that it's required logic for correct `CONFIG_PM` operation on affected SoCs and handling this with `CONFIG_PM_DEVICE` logic is out-of-scope for that, IMHO.
2. Using the notifier feature, even if modified to additionally execute before `pm_exit_post_ops`, adds runtime overhead for all returning low-power states in all drivers that make use of reinit.
My new PR #68592 will execute only when returning from the affected `power-states` and will be compiler-optimized to noop on platforms/SoCs/board configurations where it's not needed.